onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h

```diff
@@ -0,0 +1,283 @@
+// Copyright(C) 2018 Intel Corporation
+// Licensed under the MIT License
+#pragma once
+
+#include "core/graph/onnx_protobuf.h"
+#include "core/common/common.h"
+#include "core/common/exceptions.h"
+#include "core/providers/mkldnn/mkldnn_common.h"
+#include "core/providers/mkldnn/mkldnn_execution_provider.h"
+#include "core/providers/mkldnn/subgraph/mkl_kernel.h"
+#include "core/providers/mkldnn/subgraph/MklConv.h"
+#include "core/providers/mkldnn/subgraph/MklBatchNorm.h"
+#include "core/providers/mkldnn/subgraph/MklActivations.h"
+#include "core/providers/mkldnn/subgraph/MklPool.h"
+#include "core/providers/mkldnn/subgraph/MklSum.h"
+#include "core/providers/mkldnn/subgraph/MklLrn.h"
+#include "core/providers/cpu/nn/autopad_type.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+namespace {
+struct SubgraphParams {
+  std::unordered_map<std::string, ONNX_NAMESPACE::AttributeProto> attributes;
+  MKLDNNExecutionProvider* provider;
+  std::shared_ptr<Subgraph> subgraph;
+  std::shared_ptr<MKLContext> mkl_context;
+
+  SubgraphParams() {}
+};
+
+template <typename T>
+class SubgraphPrimitive : public PrimitiveBase {
+ public:
+  SubgraphPrimitive(const ONNXRunTimeTensor* input_tensors,
+                             const SubgraphParams& params)
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+
+    if (context_.net.size() == 0) {
+      CreateKernels(params);
+      Initialize(input_tensors);
+    }
+  }
+
+  void CreateKernels(const SubgraphParams& params) {
+    for (auto& mklnode : params.subgraph->mklnodes) {
+      if (mklnode.name == "Conv") {
+        std::ostringstream os;
+        os << "Conv-" << mklnode.node_index << "-";
+        std::shared_ptr<MklConv<T>> kernel;
+        kernel.reset(new MklConv<T>(mklnode, params.provider, params.mkl_context));
+        kernel->ReadAttributes(params.attributes, os.str());
```

Comment:
could populating the attributes be done as part of kernel construction? 
instead of having another public method that always gets called right after?
you can keep the method around as a private helper. but call it in the constructor?